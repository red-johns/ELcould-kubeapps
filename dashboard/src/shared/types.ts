import { IAuthState } from "../reducers/auth";
import { IServiceCatalogState } from "../reducers/catalog";
import { IConfigState } from "../reducers/config";
import { INamespaceState } from "../reducers/namespace";
import { IAppRepositoryState } from "../reducers/repos";
import { hapi } from "./hapi/release";

// Allow defining multiple error classes
// tslint:disable:max-classes-per-file

class CustomError extends Error {
  // The constructor is defined so we can later on compare the returned object
  // via err.contructor  == FOO
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
export class ForbiddenError extends CustomError {}
export class UnauthorizedError extends CustomError {}

export class NotFoundError extends CustomError {}

export class ConflictError extends CustomError {}

export class UnprocessableEntity extends CustomError {}

export interface IRepo {
  name: string;
  url: string;
}

export interface IChartVersion {
  id: string;
  attributes: IChartVersionAttributes;
  relationships: {
    chart: {
      data: IChartAttributes;
    };
  };
}

export interface IChartVersionAttributes {
  version: string;
  app_version: string;
  created: string;
}

export interface IChart {
  id: string;
  attributes: IChartAttributes;
  relationships: {
    latestChartVersion: {
      data: IChartVersionAttributes;
    };
  };
}

export interface IChartAttributes {
  name: string;
  description: string;
  home?: string;
  icon?: string;
  keywords: string[];
  maintainers: Array<{
    name: string;
    email?: string;
  }>;
  repo: IRepo;
  sources: string[];
}

export interface IChartState {
  isFetching: boolean;
  selected: {
    error?: Error;
    version?: IChartVersion;
    versions: IChartVersion[];
    readme?: string;
    readmeError?: string;
    values?: string;
  };
  items: IChart[];
}

export interface IDeployment {
  metadata: {
    name: string;
    namespace: string;
  };
}

export interface IServiceSpec {
  ports: IPort[];
  clusterIP: string;
  type: string;
}

export interface IServiceStatus {
  loadBalancer: {
    ingress?: Array<{ ip?: string; hostname?: string }>;
  };
}

export interface IPort {
  name: string;
  port: number;
  protocol: string;
  targetPort: string;
  nodePort: string;
}

export interface IHTTPIngressPath {
  path: string;
}
export interface IIngressHTTP {
  paths: IHTTPIngressPath[];
}
export interface IIngressRule {
  host: string;
  http: IIngressHTTP;
}

export interface IIngressTLS {
  hosts: string[];
}

export interface IIngressSpec {
  rules: IIngressRule[];
  tls?: IIngressTLS[];
}

export interface IResource {
  apiVersion: string;
  kind: string;
  type: string;
  spec: any;
  status: any;
  metadata: {
    name: string;
    namespace: string;
    annotations: string;
    creationTimestamp: string;
    selfLink: string;
    resourceVersion: string;
    deletionTimestamp?: string;
    uid: string;
  };
}

export interface IOwnerReference {
  apiVersion: string;
  blockOwnerDeletion: boolean;
  kind: string;
  name: string;
  uid: string;
}

export interface ISecret {
  apiVersion: string;
  kind: string;
  type: string;
  data: { [s: string]: string };
  metadata: {
    name: string;
    namespace: string;
    annotations: string;
    creationTimestamp: string;
    selfLink: string;
    resourceVersion: string;
    deletionTimestamp?: string;
    uid: string;
  };
}

export interface IDeploymentStatus {
  replicas: number;
  updatedReplicas: number;
  availableReplicas: number;
}

export interface IAppState {
  isFetching: boolean;
  error?: Error;
  deleteError?: Error;
  // currently items are always Helm releases
  items: hapi.release.Release[];
  listingAll: boolean;
  listOverview?: IAppOverview[];
  selected?: hapi.release.Release;
}

export interface IStoreState {
  catalog: IServiceCatalogState;
  apps: IAppState;
  auth: IAuthState;
  charts: IChartState;
  config: IConfigState;
  kube: IKubeState;
  repos: IAppRepositoryState;
  deployment: IDeployment;
  namespace: INamespaceState;
}

interface IK8sResource {
  apiVersion: string;
  kind: string;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#objects */
export interface IK8sObject<M, SP, ST> extends IK8sResource {
  metadata: {
    annotations?: { [key: string]: string };
    creationTimestamp?: string;
    deletionTimestamp?: string | null;
    generation?: number;
    labels?: { [key: string]: string };
    name: string;
    namespace: string;
    resourceVersion?: string;
    uid: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
  spec?: SP;
  status?: ST;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#lists-and-simple-kinds */
export interface IK8sList<I, M> extends IK8sResource {
  items: I[];
  metadata?: {
    resourceVersion?: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
}

export interface IAppRepository
  extends IK8sObject<
      {
        clusterName: string;
        creationTimestamp: string;
        deletionGracePeriodSeconds: string | null;
        deletionTimestamp: string | null;
        resourceVersion: string;
        selfLink: string;
      },
      {
        type: string;
        url: string;
        auth: {
          header: {
            secretKeyRef: {
              name: string;
              key: string;
            };
          };
        };
      },
      undefined
    > {}

export interface IAppRepositoryList
  extends IK8sList<
      IAppRepository,
      {
        continue: string;
        resourceVersion: string;
        selfLink: string;
      }
    > {}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#response-status-kind */
export interface IStatus extends IK8sResource {
  kind: "Status";
  status: "Success" | "Failure";
  message: string;
  reason:
    | "BadRequest"
    | "Unauthorized"
    | "Forbidden"
    | "NotFound"
    | "AlreadyExists"
    | "Conflict"
    | "Invalid"
    | "Timeout"
    | "ServerTimeout"
    | "MethodNotAllowed"
    | "InternalError";
  details?: {
    kind?: string;
    name?: string;
    causes?: IStatusCause[] | string;
  };
}

interface IStatusCause {
  field: string;
  message: string;
  reason: string;
}

export interface IRouterPathname {
  router: {
    location: {
      pathname: string;
    };
  };
}

export interface IRuntimeVersion {
  name: string;
  version: string;
  runtimeImage: string;
  initImage: string;
}

export interface IRuntime {
  ID: string;
  versions: IRuntimeVersion[];
  depName: string;
  fileNameSuffix: string;
}

export interface IRBACRole {
  apiGroup: string;
  namespace?: string;
  clusterWide?: boolean;
  resource: string;
  verbs: string[];
}

export interface IAppOverview {
  releaseName: string;
  namespace: string;
  version: string;
  icon?: string;
  status: string;
  chart: string;
}

export interface IKubeItem<T> {
  isFetching: boolean;
  item?: T;
  error?: Error;
}

export interface IKubeState {
  items: { [s: string]: IKubeItem<IResource> };
}
