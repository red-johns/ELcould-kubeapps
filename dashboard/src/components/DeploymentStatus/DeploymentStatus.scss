.DeploymentStatus {
  display: inline-block;
  font-size: 1em;
  line-height: 2.5em;
  padding: 0 0.875em;
  background-color: #f1f1f1;
  height: 2.5em;
}

.DeploymentStatus--success {
  color: #fff;
  background-color: #1598cb;
}

.DeploymentStatus--pending {
  background-color: #fdba12;
}

.DeploymentStatus--deleted {
  background-color: #fda5a5;
}

.DeploymentStatus--pending > .icon {
  animation: fadeinout 2s linear infinite;
}

@keyframes fadeinout {
  0%,
  100% {
    opacity: 0;
  }
  50% {
    opacity: 1;
  }
}
