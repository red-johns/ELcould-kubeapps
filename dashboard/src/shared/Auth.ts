import Axios, { AxiosResponse } from "axios";
const AuthTokenKey = "kubeapps_auth_token";

export class Auth {
  public static getAuthToken() {
    return localStorage.getItem(AuthTokenKey);
  }

  public static setAuthToken(token: string) {
    return localStorage.setItem(AuthTokenKey, token);
  }

  public static unsetAuthToken() {
    return localStorage.removeItem(AuthTokenKey);
  }

  public static wsProtocols() {
    const token = this.getAuthToken();
    if (!token) {
      return [];
    }
    return [
      "base64url.bearer.authorization.k8s.io." + btoa(token).replace(/=*$/g, ""),
      "binary.k8s.io",
    ];
  }

  public static fetchOptions(): RequestInit {
    const headers = new Headers();
    headers.append("Authorization", `Bearer ${this.getAuthToken()}`);
    return {
      headers,
    };
  }

  // Throws an error if the token is invalid
  public static async validateToken(token: string) {
    try {
      await Axios.get("api/kube/", { headers: { Authorization: `Bearer ${token}` } });
    } catch (e) {
      const res = e.response as AxiosResponse;
      if (res.status === 401) {
        throw new Error("invalid token");
      }
    }
  }
}

export const axios = Axios.create();
