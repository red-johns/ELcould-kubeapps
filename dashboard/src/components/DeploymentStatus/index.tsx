import DeploymentStatus from "./DeploymentStatus";

export default DeploymentStatus;
