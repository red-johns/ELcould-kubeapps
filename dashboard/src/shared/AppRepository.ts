import { axios } from "./Auth";
import { IAppRepository, IAppRepositoryList } from "./types";

export class AppRepository {
  public static async list(namespace: string) {
    const { data } = await axios.get<IAppRepositoryList>(AppRepository.getResourceLink(namespace));
    return data;
  }

  public static async get(name: string, namespace: string) {
    const { data } = await axios.get(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async update(name: string, namespace: string, newApp: IAppRepository) {
    const { data } = await axios.put(AppRepository.getSelfLink(name, namespace), newApp);
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const { data } = await axios.delete(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async create(name: string, namespace: string, url: string, auth: any) {
    const { data } = await axios.post<IAppRepository>(AppRepository.getResourceLink(namespace), {
      apiVersion: "kubeapps.com/v1alpha1",
      kind: "AppRepository",
      metadata: {
        name,
      },
      spec: { auth, type: "helm", url },
    });
    return data;
  }

  private static APIBase: string = "api/kube";
  private static APIEndpoint: string = `${AppRepository.APIBase}/apis/kubeapps.com/v1alpha1`;
  private static getResourceLink(namespace?: string): string {
    return `${AppRepository.APIEndpoint}/${
      namespace ? `namespaces/${namespace}/` : ""
    }apprepositories`;
  }
  private static getSelfLink(name: string, namespace: string): string {
    return `${AppRepository.APIEndpoint}/namespaces/${namespace}/apprepositories/${name}`;
  }
}
