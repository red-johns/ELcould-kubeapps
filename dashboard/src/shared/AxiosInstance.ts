import { AxiosError, AxiosInstance, AxiosRequestConfig } from "axios";
import { Action, Store } from "redux";
import { ThunkDispatch } from "redux-thunk";
import actions from "../actions";
import { Auth } from "./Auth";
import {
  ConflictError,
  ForbiddenError,
  IStoreState,
  NotFoundError,
  UnauthorizedError,
  UnprocessableEntity,
} from "./types";

// createAxiosInterceptors will configure a set of interceptors to a provided axios instance,
// relying also on an external redux store for action dispatching
export function createAxiosInterceptors(axios: AxiosInstance, store: Store<IStoreState>) {
  axios.interceptors.request.use((config: AxiosRequestConfig) => {
    const authToken = Auth.getAuthToken();
    if (authToken) {
      config.headers.Authorization = `Bearer ${authToken}`;
    }
    return config;
  });
  axios.interceptors.response.use(
    response => response,
    e => {
      const err: AxiosError = e;
      let message = err.message;
      if (err.response && err.response.data.message) {
        message = err.response.data.message;
      }
      switch (err.response && err.response.status) {
        case 401:
          // Global action dispatch to log the user out
          const dispatch = store.dispatch as ThunkDispatch<IStoreState, null, Action>;
          dispatch(actions.auth.authenticationError(message));
          dispatch(actions.auth.logout());
          return Promise.reject(new UnauthorizedError(message));
        case 403:
          return Promise.reject(new ForbiddenError(message));
        case 404:
          return Promise.reject(new NotFoundError(message));
        case 409:
          return Promise.reject(new ConflictError(message));
        case 422:
          return Promise.reject(new UnprocessableEntity(message));
        default:
          return Promise.reject(new Error(message));
      }
    },
  );
}
