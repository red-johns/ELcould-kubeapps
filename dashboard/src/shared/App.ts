import { AppRepository } from "./AppRepository";
import { axios } from "./Auth";
import { hapi } from "./hapi/release";
import { IAppOverview, IChartVersion } from "./types";

export const TILLER_PROXY_ROOT_URL = "api/tiller-deploy/v1";

export class App {
  public static getResourceURL(namespace?: string, name?: string, query?: string) {
    let url = TILLER_PROXY_ROOT_URL;
    if (namespace) {
      url += `/namespaces/${namespace}`;
    }
    url += "/releases";
    if (name) {
      url += `/${name}`;
    }
    if (query) {
      url += `?${query}`;
    }
    return url;
  }

  public static async create(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const repo = await AppRepository.get(chartAttrs.repo.name, kubeappsNamespace);
    const auth = repo.spec.auth;
    const endpoint = App.getResourceURL(namespace);
    const { data } = await axios.post(endpoint, {
      auth,
      chartName: chartAttrs.name,
      releaseName,
      repoUrl: chartAttrs.repo.url,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const repo = await AppRepository.get(chartAttrs.repo.name, kubeappsNamespace);
    const auth = repo.spec.auth;
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axios.put(endpoint, {
      auth,
      chartName: chartAttrs.name,
      releaseName,
      repoUrl: chartAttrs.repo.url,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async delete(releaseName: string, namespace: string, purge: boolean) {
    let purgeQuery;
    if (purge) {
      purgeQuery = "purge=true";
    }
    const { data } = await axios.delete(App.getResourceURL(namespace, releaseName, purgeQuery));
    return data;
  }

  public static async listApps(namespace?: string, allStatuses?: boolean) {
    let query;
    if (allStatuses) {
      query = "statuses=all";
    }
    const { data } = await axios.get<{ data: IAppOverview[] }>(
      App.getResourceURL(namespace, undefined, query),
    );
    return data.data;
  }

  public static async getRelease(namespace: string, name: string) {
    const { data } = await axios.get<{ data: hapi.release.Release }>(
      this.getResourceURL(namespace, name),
    );
    return data.data;
  }
}
