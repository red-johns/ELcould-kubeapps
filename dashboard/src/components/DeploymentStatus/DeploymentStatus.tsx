import * as React from "react";

import { AlertTriangle } from "react-feather";
import isSomeResourceLoading from "../../components/AppView/helpers";
import Check from "../../icons/Check";
import Compass from "../../icons/Compass";
import { hapi } from "../../shared/hapi/release";
import { IDeploymentStatus, IKubeItem, IResource } from "../../shared/types";
import "./DeploymentStatus.css";

interface IDeploymentStatusProps {
  deployments: Array<IKubeItem<IResource>>;
  info?: hapi.release.IInfo;
}

class DeploymentStatus extends React.Component<IDeploymentStatusProps> {
  public render() {
    if (isSomeResourceLoading(this.props.deployments)) {
      return <span className="DeploymentStatus">Loading...</span>;
    }
    if (this.props.info && this.props.info.deleted) {
      return this.renderDeletedStatus();
    }
    return this.isReady() ? this.renderSuccessStatus() : this.renderPendingStatus();
  }

  private renderSuccessStatus() {
    return (
      <span className="DeploymentStatus DeploymentStatus--success">
        <Check className="icon padding-t-tiny" /> Deployed
      </span>
    );
  }

  private renderPendingStatus() {
    return (
      <span className="DeploymentStatus DeploymentStatus--pending">
        <Compass className="icon padding-t-tiny" /> Deploying
      </span>
    );
  }

  private isReady() {
    const { deployments } = this.props;
    if (deployments.length > 0) {
      // Check if all the deployments has the same number of
      // desired and available replicas.
      return deployments.every(d => {
        if (d.item) {
          const status: IDeploymentStatus = d.item.status;
          return status.availableReplicas === status.replicas;
        }
        return false;
      });
    } else {
      // if there are no deployments, then the app is considered "ready"
      // TODO: this currently does not distinguish between deployments not
      // loaded yet and no deployments
      return true;
    }
  }

  private renderDeletedStatus() {
    return (
      <span className="DeploymentStatus DeploymentStatus--deleted">
        <AlertTriangle className="icon" style={{ bottom: "-0.425em" }} /> Deleted
      </span>
    );
  }
}

export default DeploymentStatus;
