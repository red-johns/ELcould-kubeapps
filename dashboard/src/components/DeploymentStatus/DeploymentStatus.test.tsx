import { shallow } from "enzyme";
import * as React from "react";

import { IResource } from "shared/types";
import DeploymentStatus from "./DeploymentStatus";

it("renders a loading status", () => {
  const deployments = [
    {
      isFetching: true,
    },
  ];
  const wrapper = shallow(<DeploymentStatus deployments={deployments} />);
  expect(wrapper.text()).toContain("Loading");
  expect(wrapper).toMatchSnapshot();
});

it("renders a deleting status", () => {
  const deployments = [
    {
      isFetching: false,
    },
  ];
  const wrapper = shallow(<DeploymentStatus deployments={deployments} info={{ deleted: {} }} />);
  expect(wrapper.text()).toContain("Deleted");
  expect(wrapper).toMatchSnapshot();
});

it("renders a deploying status", () => {
  const deployments = [
    {
      isFetching: false,
      item: {
        status: {
          replicas: 1,
          availableReplicas: 0,
        },
      } as IResource,
    },
  ];
  const wrapper = shallow(<DeploymentStatus deployments={deployments} />);
  expect(wrapper.text()).toContain("Deploying");
  expect(wrapper).toMatchSnapshot();
});

it("renders a deployed status", () => {
  const deployments = [
    {
      isFetching: false,
      item: {
        status: {
          replicas: 1,
          availableReplicas: 1,
        },
      } as IResource,
    },
  ];
  const wrapper = shallow(<DeploymentStatus deployments={deployments} />);
  expect(wrapper.text()).toContain("Deployed");
  expect(wrapper).toMatchSnapshot();
});
