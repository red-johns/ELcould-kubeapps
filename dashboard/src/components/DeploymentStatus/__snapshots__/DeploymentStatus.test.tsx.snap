// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a deleting status 1`] = `
<span
  className="DeploymentStatus DeploymentStatus--deleted"
>
  <AlertTriangle
    className="icon"
    color="currentColor"
    size="24"
    style={
      Object {
        "bottom": "-0.425em",
      }
    }
  />
   Deleted
</span>
`;

exports[`renders a deployed status 1`] = `
<span
  className="DeploymentStatus DeploymentStatus--success"
>
  <Check
    className="icon padding-t-tiny"
  />
   Deployed
</span>
`;

exports[`renders a deploying status 1`] = `
<span
  className="DeploymentStatus DeploymentStatus--pending"
>
  <Compass
    className="icon padding-t-tiny"
  />
   Deploying
</span>
`;

exports[`renders a loading status 1`] = `
<span
  className="DeploymentStatus"
>
  Loading...
</span>
`;
